File: rxretrofit/src/main/java/com/fancy/rxretrofit/base/BaseCallBack.java
package com.fancy.rxretrofit.base;

import android.content.Context;
import android.support.annotation.NonNull;

import com.fancy.rxretrofit.error.Error;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * retrofit基本返回值
 *
 * @author fanlei
 * @version 1.0 2019\1\4 0004
 * @since JDK 1.7
 */
public abstract class BaseCallBack<T> implements Callback<T> {

    protected final int defaultFailCode = -1;

    protected final Context mContext;

    public BaseCallBack(@NonNull Context context) {
        this.mContext = context;
    }


    @Override
    public void onResponse(Call<T> call, Response<T> response) {
        if (response.raw().code() == 200) {
            onSuc(response);
        } else {//失败响应
            onFail(response, Error.buildError(response).getErrorMessage(), response.raw().code());
        }

    }

    @Override
    public void onFailure(Call<T> call, Throwable t) {//网络问题会走该回调
        onFail(null, Error.buildError(t).getErrorMessage(), defaultFailCode);
    }

    public abstract void onSuc(Response<T> response);

    public abstract void onFail(Response<T> response, String message, int failCode);

}

File: rxretrofit/src/main/java/com/fancy/rxretrofit/callback/BaseCallBack.kt
package com.fancy.rxretrofit.callback

import android.content.Context
import com.fancy.rxretrofit.NetError
import com.fancy.rxretrofit.impl.LoadingListener
import com.fancy.rxretrofit.load.Gloading
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 *
 * @Description:
 * @Author: fanlei
 * @CreateDate: 13/9/2021 下午3:37
 * @Version: 1.0
 */
abstract class BaseCallBack<T>(context: Context) : Callback<T>, LoadingListener {
    protected val defaultFailCode = -1

    protected var mContext: Context? = null
    private var holder: Gloading.Holder? = null

    init {
        mContext = context
    }

    override fun onResponse(call: Call<T>?, response: Response<T>) {
        holder?.showLoadSuccess()
        if (response.raw().code() == 200) {
            onObtain(response)
        } else { //失败响应
            onLose(
                    response,
                    NetError.buildError(response)!!.getErrorMessage(),
                    response.raw().code()
            )
        }
    }

    override fun onFailure(call: Call<T>?, t: Throwable?) { //网络问题会走该回调
        holder?.showLoadSuccess()
        onLose(null, NetError.buildError(t!!)!!.getErrorMessage(), defaultFailCode)
    }


    protected abstract fun onObtain(response: Response<T>?)

    protected abstract fun onLose(response: Response<T>?, message: String?, failCode: Int)

    override fun setLoading(holder: Gloading.Holder?) {
        this.holder = holder
    }
}